### PR TITLE
feat: Carousel UI component

### PR DESCRIPTION
**Pull request by liamcho**
Fixes: [AC-1230](https://sendbird.atlassian.net/browse/AC-1230)

### Changelogs

- Added `Carousel` ui component
- `MessageTemplate` now supports composite templates

[AC-1230]: https://sendbird.atlassian.net/browse/AC-1230?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

reference: https://stackblitz.com/edit/stackblitz-starters-bmuznq?file=src%2FCarousel.tsx,src%2FApp.tsx

### CODE REVIEW

File: src/ui/TemplateMessageItemBody/index.tsx [hunk 1]
```diff
@@ -171,10 +180,15 @@ export function TemplateMessageItemBody({
           );
           return filledMessageTemplateItems;
         });
+      setShowLoading(false);
       setFilledMessageTemplateItemsList(filledMessageTemplateItemsList);
     }
   }, [templateData.key, waitingTemplateKeysMapString]);
 
+  useEffect(() => {
+    onMessageHeightChange();
```
[the remark this answers is not in the document]
Comment (liamcho) [reply src/ui/TemplateMessageItemBody/index.tsx:193]: Screenshot 2024-03-12 at 12 45 41 PM
tried this but not working.

[the remark this answers is not in the document]
Comment (liamcho) [reply src/ui/TemplateMessageItemBody/index.tsx:193]: Screenshot 2024-03-12 at 12 49 43 PM
Screenshot 2024-03-12 at 12 49 53 PM

Looking into alternative solution

File: src/lib/dux/appInfo/reducers.ts [hunk 2]
```diff
@@ -13,39 +13,44 @@ export default function reducer(state: AppInfoStateType, action: AppInfoActionTy
...
-          isError: false,
-        };
+        const { keys, requestedAt } = payload;
+        keys.forEach((key) => {
+          state.waitingTemplateKeysMap[key] = {
+            erroredMessageIds: state.waitingTemplateKeysMap[key]?.erroredMessageIds ?? [],
+            requestedAt,
+          };
+        });
         return { ...state };
```
Comment (HoonBaek) [inline src/lib/dux/appInfo/reducers.ts:42]: Same here, for example we can do like this
```tsx
const { keys, requestedAt } = payload;
const updatedWaitingTemplateKeysMap = { ...state.waitingTemplateKeysMap };
keys.forEach((key) => {
  updatedWaitingTemplateKeysMap[key] = {
    erroredMessageIds: updatedWaitingTemplateKeysMap[key]?.erroredMessageIds ?? [],
    requestedAt,
  };
});
return {
  ...state,
  waitingTemplateKeysMap: updatedWaitingTemplateKeysMap,
};
```

File: src/lib/dux/appInfo/reducers.ts [hunk 3]
```diff
@@ -13,39 +13,44 @@ export default function reducer(state: AppInfoStateType, action: AppInfoActionTy
...
-        if (waitingTemplateKeyData) {
-          waitingTemplateKeyData.isError = true;
-        }
+        const { keys, messageId } = payload;
+        keys.forEach((key) => {
+          const waitingTemplateKeyData: WaitingTemplateKeyData | undefined = state.waitingTemplateKeysMap[key];
+          if (waitingTemplateKeyData && waitingTemplateKeyData.erroredMessageIds.indexOf(messageId) === -1) {
+            waitingTemplateKeyData.erroredMessageIds.push(messageId);
+          }
+        });
```
Comment (HoonBaek) [inline src/lib/dux/appInfo/reducers.ts:53]: same here

File: src/lib/hooks/useMessageTemplateUtils.ts [hunk 4]
```diff
@@ -128,54 +128,72 @@ export default function useMessageTemplateUtils({
...
-      if (parsedTemplate) {
+      if (newParsedTemplates.length > 0) {
         // Update cache
         const cachedMessageTemplates: string | null = localStorage.getItem(CACHED_MESSAGE_TEMPLATES_KEY);
         if (cachedMessageTemplates) {
           const parsedTemplates: SendbirdMessageTemplate[] = JSON.parse(cachedMessageTemplates);
-          parsedTemplates.push(parsedTemplate);
+          const existingKeys = parsedTemplates.map((parsedTemplate) => parsedTemplate.key);
+          newParsedTemplates.forEach((newParsedTemplate) => {
+            if (existingKeys.indexOf(newParsedTemplate.key) === -1) {
```
Comment (HoonBaek) [inline src/lib/hooks/useMessageTemplateUtils.ts:168]: How about this?
```suggestion
            if (!existingKeys.includes(newParsedTemplate.key)) {
```

File: src/modules/GroupChannel/index.tsx [hunk 5]
```diff
@@ -5,6 +5,7 @@ import GroupChannelUI, { GroupChannelUIProps } from './components/GroupChannelUI
 
 export interface GroupChannelProps extends GroupChannelProviderProps, GroupChannelUIProps {}
 export const GroupChannel = (props: GroupChannelProps) => {
+  console.log('## channelurl: ', props.channelUrl);
```
Comment (HoonBaek) [inline src/modules/GroupChannel/index.tsx:8]: It seems need to remove this before merge

File: src/ui/Carousel/index.tsx [hunk 6]
```diff
@@ -0,0 +1,258 @@
+import './index.scss';
+import React, {ReactElement, useEffect, useRef, useState} from 'react';
```
Comment (HoonBaek) [inline src/ui/Carousel/index.tsx:2]: nit: need to run lint

File: src/ui/Carousel/index.tsx [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+interface CarouselItemProps {
+  key: string;
+  item: ReactElement;
```
Comment (HoonBaek) [inline src/ui/Carousel/index.tsx:6]: Is it certain that the type of this is `ReactElement`?
Don't we need to define a separate type for this item?

Comment (liamcho) [reply src/ui/Carousel/index.tsx:17]: so the item is 
```
export const MessageTemplateWrapper = ({
  message,
  templateItems,
}: MessageTemplateWrapperProps): ReactElement => {
  return <MessageTemplateProvider message={message}>
    <MessageTemplate templateItems={templateItems} />
  </MessageTemplateProvider>;
};
```

but Im not sure if it is 
`(alias) const MessageTemplateProvider: (props: MessageContextProps) => React.JSX.Element`
or
`(alias) function MessageTemplate({ templateItems }: MessageTemplateProps): React.JSX.Element`

File: src/ui/Carousel/index.tsx [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+  end: number;
+}
+
+interface CarouselItemProps {
+  key: string;
+  item: ReactElement;
+  defaultWidth: string;
+}
+
+function shouldRenderAsFixed(item: ReactElement) {
```
Comment (bang9) [inline src/ui/Carousel/index.tsx:21]: 왜 fixed 로 렌더링해야 하는지에 대한 코멘트 하나 남겨주시면 좋을것 같습니다.

**Review by bang9 — APPROVED** [review]

**Review by HoonBaek — APPROVED** [review]
Nice!

File: src/ui/Carousel/index.tsx (hunk 8 above)
Comment (liamcho) [reply src/ui/Carousel/index.tsx:26]: 완료요!

**Pull request merged**